src/mixins/itext_behavior.mixin.ts

```diff
@@ -531,67 +532,70 @@ export abstract class ITextBehaviorMixin<
-    data: {
+    {
+      selectionStart,
+      selectionEnd,
+    }: {
       selectionStart: number;
       selectionEnd: number;
-      text: string;
-      value: string;
     }
   ) {
-    const t = this.calcTransformMatrix();
     const flipFactor = new Point(this.flipX ? -1 : 1, this.flipY ? -1 : 1);
-    const boundaries = this._getCursorBoundaries(data.selectionStart);
+    const boundaries = this._getCursorBoundaries(selectionStart);
     const selectionPosition = new Point(
       boundaries.left + boundaries.leftOffset,
       boundaries.top + boundaries.topOffset
     ).multiply(flipFactor);
-    const pos = transformPoint(selectionPosition, t);
+    const pos = selectionPosition.transform(this.calcTransformMatrix());
     const pointer = this.canvas.getPointer(e);
     const diff = pointer.subtract(pos);
     const enableRetinaScaling = this.canvas._isRetinaScaling();
     const retinaScaling = this.getCanvasRetinaScaling();
     const bbox = this.getBoundingRect(true);
     const correction = pos.subtract(new Point(bbox.left, bbox.top));
-    const offset = correction.add(diff).scalarMultiply(retinaScaling);
+    const vpt = this.canvas.viewportTransform;
+    const offset = correction.add(diff).transform(vpt, true);
     //  prepare instance for drag image snapshot by making all non selected text invisible
     const bgc = this.backgroundColor;
-    const styles = object.clone(this.styles, true);
+    const styles = clone(this.styles, true);
     delete this.backgroundColor;
     const styleOverride = {
+      stroke: 'transparent',
       fill: 'transparent',
       textBackgroundColor: 'transparent',
     };
-    this.setSelectionStyles(styleOverride, 0, data.selectionStart);
-    this.setSelectionStyles(styleOverride, data.selectionEnd, data.text.length);
+    this.setSelectionStyles(styleOverride, 0, selectionStart);
+    this.setSelectionStyles(styleOverride, selectionEnd, this.text.length);
     let dragImage = this.toCanvasElement({
-      enableRetinaScaling: enableRetinaScaling,
+      enableRetinaScaling,
     });
+    // restore values
     this.backgroundColor = bgc;
     this.styles = styles;
-    //  handle retina scaling
-    if (enableRetinaScaling && retinaScaling > 1) {
-      const c = createCanvasElement();
-      c.width = dragImage.width / retinaScaling;
-      c.height = dragImage.height / retinaScaling;
-      const ctx = c.getContext('2d');
+    //  handle retina scaling and vpt
+    if (retinaScaling > 1 || !isIdentityMatrix(vpt)) {
+      const dragImageCanvas = createCanvasElement();
+      const size = new Point(dragImage.width, dragImage.height)
+        .scalarDivide(retinaScaling)
+        .transform(vpt, true);
+      dragImageCanvas.width = size.x;
+      dragImageCanvas.height = size.y;
+      const ctx = dragImageCanvas.getContext('2d');
       ctx.scale(1 / retinaScaling, 1 / retinaScaling);
+      const [a, b, c, d] = vpt;
+      const origin = new Point().transform(vpt, true);
```

Comment:
isn't this exactly vpt[4] and vpt[5]? what is this line doing?

Reply:
you might be right.
I will check.

Reply:
the logic behind it is the same as send 0,0 to the transformed plane